Create network elements

EtienneLt opened the pull request:
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

Feature.

**What is the current behavior?** *(You can also link to an open issue here)*

Network elements cannot be created through python methods.

**What is the new behavior (if this is a feature change)?**

We can now create network elements, for example: 
```python
def test_create_network_and_run_loadflow():
    n = pn.create_empty()
    stations = pd.DataFrame.from_records(index='id', data=[
        {'id': 'S1', 'country': 'BE'},
        {'id': 'S2', 'country': 'DE'}
    ])
    n.create_substations(stations)

    voltage_levels = pd.DataFrame.from_records(index='id', data=[
        {'substation_id': 'S1', 'id': 'VL1', 'topology_kind': 'BUS_BREAKER', 'nominal_v': 400},
        {'substation_id': 'S2', 'id': 'VL2', 'topology_kind': 'BUS_BREAKER', 'nominal_v': 400},
    ])
    n.create_voltage_levels(voltage_levels)

    buses = pd.DataFrame.from_records(index='id', data=[
        {'voltage_level_id': 'VL1', 'id': 'B1'},
        {'voltage_level_id': 'VL2', 'id': 'B2'},
    ])
    n.create_buses(buses)

    lines = pd.DataFrame.from_records(index='id', data=[
        {'id': 'LINE', 'voltage_level1_id': 'VL1', 'voltage_level2_id': 'VL2', 'bus1_id': 'B1', 'bus2_id': 'B2',
         'r': 0.1, 'x': 1.0, 'g1': 0, 'b1': 1e-6, 'g2': 0, 'b2': 1e-6}
    ])
    n.create_lines(lines)

    loads = pd.DataFrame.from_records(index='id', data=[
        {'voltage_level_id': 'VL2', 'id': 'LOAD', 'bus_id': 'B2', 'p0': 100, 'q0': 10}
    ])
    n.create_loads(loads)

    generators = pd.DataFrame.from_records(index='id', data=[
        {'voltage_level_id': 'VL1', 'id': 'GEN', 'bus_id': 'B1', 'target_p': 100, 'min_p': 0, 'max_p': 200,
         'target_v': 400, 'voltage_regulator_on': True}
    ])
    n.create_generators(generators)

    import pypowsybl.loadflow as lf
    lf.run_ac(n)

    line = n.get_lines().loc['LINE']
    assert line.p2 == pytest.approx(-100, abs=1e-2)
    assert line.q2 == pytest.approx(-10, abs=1e-2)
    assert line.p1 == pytest.approx(100, abs=1e-1)
    assert line.q1 == pytest.approx(9.7, abs=1e-1)
```

EtienneLt committed:
manage severals indexes

Signed-off-by: Etienne LESOT <etienne.lesot@rte-france.com>

EtienneLt committed:
add update for multi indexes and implements it for shunt non linear

Signed-off-by: Etienne LESOT <etienne.lesot@rte-france.com>

EtienneLt committed:
add create network elements

Signed-off-by: Etienne LESOT <etienne.lesot@rte-france.com>

EtienneLt committed:
refactor c code on updating dataframe

Signed-off-by: Etienne LESOT <etienne.lesot@rte-france.com>

annetill commented:
@EtienneLt it seems that your PR is needed for long-term studies. Do you think that you can rework on that in the next following months? It would be great!

EtienneLt commented:
> 

yes no problem

sylvlecl committed:
Merge branch 'main' into create-network-equipment

sylvlecl committed:
WIP

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
WIP

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
WIP: adapt to main

- expose list of series metadata for element creation dataframes
- isolate elements creation tests
- add some missing creation (HVDC lines, substations)

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Add substations

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Add HVDC lines and switches creation + some fixes

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
WIP: list of metadata

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
More strongly typed C interface

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
fix

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Non linear shunts handling

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Support for tap changers

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Add missing licenses

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Some renaming

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Remove obsolete code

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix unit test

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Stronger typing of C API

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'main' into create-network-equipment

sylvlecl committed:
Fix build

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'main' into create-network-equipment

sylvlecl committed:
Add index metadata to support unnamed indexes

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Support for keyword arguments

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Add exception tests

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl commented:
Actually something is missing for node breaker topologies:
we cannot create "internal connections". Still a todo.

sylvlecl committed:
Add full node breaker unit test

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Provide expected attributes in docstring

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge branch 'main' into create-network-equipment

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix typing

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Some linting

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/BatteryDataframeAdder.java:
```diff
@@ -0,0 +1,55 @@
+        BatteryAdder batteryAdder = network.getVoltageLevel(dataframe.getStringValue("voltage_level_id", indexElement)
+                        .orElseThrow(() -> new PowsyblException("voltage_level_id is missing")))
+                .newBattery();
+        NetworkElementCreationUtils.createInjection(batteryAdder, dataframe, indexElement);
```
Be careful here: if not present, I think that you will have NaN values.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/BatteryDataframeAdder.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static final List<SeriesMetadata> METADATA = List.of(
+            SeriesMetadata.stringIndex("id"),
+            SeriesMetadata.strings("voltage_level_id"),
```
Just a general question: when you only fill the bus id, and not the connectable bus id, what is happening ? And if you want a disconnected battery, you will have to fill the connectable bus only ?

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/DanglingLineDataframeAdder.java:
```diff
@@ -0,0 +1,58 @@
+    public void addElement(Network network, UpdatingDataframe dataframe, int indexElement) {
+        DanglingLineAdder adder = network.getVoltageLevel(dataframe.getStringValue("voltage_level_id", indexElement)
+                .orElseThrow(() -> new PowsyblException("voltage_level_id is missing"))).newDanglingLine();
+        NetworkElementCreationUtils.createInjection(adder, dataframe, indexElement);
```
I think that we have to think about `g` and a `b` default values equals to zero.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/GeneratorDataframeAdder.java:
```diff
@@ -0,0 +1,62 @@
+        GeneratorAdder generatorAdder = network.getVoltageLevel(dataframe.getStringValue("voltage_level_id", indexElement)
+                        .orElseThrow(() -> new PowsyblException("voltage_level_id is missing")))
+                .newGenerator();
+        NetworkElementCreationUtils.createInjection(generatorAdder, dataframe, indexElement);
```
I think that we have to think about default values for minP and maxP (it seems to be NaN in powsybl-core), right?

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/HvdcDataframeAdder.java:
```diff
@@ -0,0 +1,55 @@
+        NetworkElementCreationUtils.createIdentifiable(adder, dataframe, indexElement);
+        dataframe.getStringValue("converter_station1_id", indexElement).ifPresent(adder::setConverterStationId1);
+        dataframe.getStringValue("converter_station2_id", indexElement).ifPresent(adder::setConverterStationId2);
+        dataframe.getDoubleValue("max_p", indexElement).ifPresent(adder::setMaxP);
```
Default value for maxP ?

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/LineDataframeAdder.java:
```diff
@@ -0,0 +1,60 @@
+    public void addElement(Network network, UpdatingDataframe dataframe, int indexElement) {
+        LineAdder lineAdder = network.newLine();
+        NetworkElementCreationUtils.createBranch(lineAdder, dataframe, indexElement);
+        dataframe.getDoubleValue("b1", indexElement).ifPresent(lineAdder::setB1);
```
I think that for line, we must think about by default having b1, b2, g1 and g2 equal to zero.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/PhaseTapChangerDataframeAdder.java:
```diff
@@ -0,0 +1,76 @@
+    private static final List<SeriesMetadata> METADATA = List.of(
+            SeriesMetadata.stringIndex("id"),
+            SeriesMetadata.strings("regulation_mode"),
+            SeriesMetadata.doubles("target_deadband"),
```
The target deadband could be at zero by default to ease understanding.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/PhaseTapChangerDataframeAdder.java:
```diff
@@ -0,0 +1,76 @@
+            SeriesMetadata.ints("tap")
+    );
+
+    private static final List<SeriesMetadata> STEPS_METADATA = List.of(
```
For phase tap changers, I think that we can provide a lot of default values and maybe allow only angles to be set.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/RatioTapChangerDataframeAdder.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class RatioTapChangerDataframeAdder implements NetworkElementAdder {
+
+    private static final List<SeriesMetadata> METADATA = List.of(
```
On load can be to true by default, and the target deadband could be to zero. For the ratio tap changer, I think that it could be great to have the possibility to set only ratios.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/ShuntDataframeAdder.java:
```diff
@@ -0,0 +1,108 @@
+ * @author Sylvain Leclerc <sylvain.leclerc@rte-france.com>
+ */
+public class ShuntDataframeAdder implements NetworkElementAdder {
+
```
In general, for shunt, it could be great to have a way to only set a B from a dataframe, that will create a shunt understandable by Powsybl.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/SwitchDataframeAdder.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class SwitchDataframeAdder extends AbstractSimpleAdder {
+
+    private static final List<SeriesMetadata> METADATA = List.of(
```
In general: it always has been strange to be to give both voltage level id and bus id. It will be very great to have the possibility to give only busId to create equipments in the network, and to have a way to retrieve the voltage level. It is a bit confusing to expose the complexity of IIDM in python.

sylvlecl replied:
See discussion synthesis about default values in comment below.

annetill commented on java/src/main/java/com/powsybl/dataframe/network/adders/TwtDataframeAdder.java:
```diff
@@ -0,0 +1,84 @@
+            SeriesMetadata.doubles("rated_u1"),
+            SeriesMetadata.doubles("rated_u2"),
+            SeriesMetadata.doubles("rated_s"),
+            SeriesMetadata.doubles("b"),
```
Same remark as lines. And maybe rated U1 and U2 could also have default values equal to nominal voltage at ends. What do you think ?

sylvlecl replied:
See discussion synthesis about default values in comment below.

sylvlecl commented:
Just writing down the outcome of the discussions about **default values**:

currently, we inherit the behaviour of powsybl-core adders.
The implementation of default values will be more valuable in powsybl-core than in pypowsybl, because it will benefit to all users.

Therefore, unless there is something specific that the python community asks for, and that we really don't want to implement in powsybl-core, we'll keep it this way.

sylvlecl committed:
Merge branch 'main' into create-network-equipment

sylvlecl committed:
Adapt elements selection to C structure changes

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix typing

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

annetill reviewed (approved):
I am going to try to put some default values in the adders of powsybl-core, and we will see what remains for pypowsybl !
